src/main/java/baritone/behavior/PathingBehavior.java

```diff
@@ -99,6 +103,11 @@ public void onTick(TickEvent event) {
             baritone.getPathingControlManager().cancelEverything();
             return;
         }
+
+        if (ticksElapsedSoFar % 200 == 0) {
+            precomputedData = new PrecomputedData(); // This is here for now in case settings aren't changed in normal ways, should mean it is updated whatever once every 10 seconds
+        }
```

Comment:
Yes, that's about what I meant.
My main concern about this is settings being outdated for a long enough time for users to notice it and we also can't speed this up to every tick like I though when writing my previous comment because now that I looked at this pr again I noticed `PrecomputedDataForBlockState` eagerly computes all data on construction, meaning any approach like this results in a repeated waste of cpu time. The two solutions I see are
1. making `PrecomputedDataForBlockState` lazy.
This would probably require an additional `boolean[]` and an additional check, resulting in more memory and time consumption on the pathing thread.
2. Not reconstructing `PrecomputedData` on a regular schedule.
Since reconstructing it every 10 seconds already makes it very likely that it is not reused for a second path, why not stop trying to do so and reconstruct it once per path?
Alternatively we could cache the (relevant) setting values and only reconstruct `PrecomputedData` if there was an actual change. This is the option I had in mind but thought it's too much effort for no gain over the simpler solution you implemented. This can also be combined with 2.

Reply:
I was initially going to do it whenever a new CalculationContext is created, however it seems a new one is created every tick, so where would you propose that we update this?

Reply:
ZacSharp makes a good point, I didn't conisder that: https://github.com/cabaletta/baritone/pull/3479#discussion_r890597059 

Walking through all BlockStates already takes a nontrivial amount of time, there are thousands of them, scorbett mentioned in the OP that precomputing all the data takes 7 to 10ms. That's very nontrivial, if we did that on the main game thread it could cause noticable stutter. For example, at 60fps, ticks can't take longer than 16.666ms otherwise the framerate will drop. With just a few mods adding new blockstates at 60fps, or with a higher than 60fps monitor with vanilla, this could cause notable performance issues. This would definitely not be desirable on the main thread.

So I think that, for sure, we **should not** precompute all block states on the main tick thread, at any time interval (whether it's once a tick, or once every 200 ticks).

Precomputing every time a setting is modified would be nice, but, as ZacSharp points out, other mods and clients can mess with Baritone's settings at any time, and we don't necessarily get a callback.


There are a whole bunch of options here with a lot of tradeoffs, here are a few:

- Fully lazy `PrecomputedDataForBlockState`. This would be relatively simple, but it would add an extra branch to every lookup, which is exactly what this is trying to avoid. This is a perfectly reasonable option and it would still be much better than before. Note that `ToolSet` already does exactly this with its HashMap.
- Construct the precomputed data on the pathing thread, first thing before it starts A*. This would work and it wouldn't make the main game stutter. It would be less than ideal from a thread safety pov, but, this is what Baritone does today (accessing settings from pathing thread), so, nothing new. We have a 50ms step function for pathing, which is that the pathing thread's completed path can't be presented to the game until the beginning of the next tick when `PathingBehavior` checks for it. That means that it's more or less "fine" to add a short delay to its startup. For example, even for tiny paths such as when using `MineProcess`, this wouldn't change anything, it would still be a 1 tick delay, it's just that instead of 5ms of the 50ms budget being used, we now use 15ms or something. However, this could still pose a problem with modpacks that add more blockstates, or, future versions of minecraft that do the same. Also not everyone has a fast computer.
- Some kind of godawful reflect to check when a Setting value has changed. Some kind of horribleness with `System.identityHashCode` should work? This could be made a little better by only checking the settings that are actually referenced by `canWalkThrough` and `canWalkOn` maybe? For the `Setting<List<?>>` ones, maybe we could make the `List` be a custom `TattleTaleList` that intercepts calls to `add` and `remove` and tattletales by telling the precomputed data to recompute itself? This is probably more trouble than it's worth.
- Precompute the data on a regular schedule, but on a totally separate thread from both main game and pathing? e.g. just `while (true) { Baritone.precomputedData = compute(); Thread.sleep(10000); }`
- Some kind of hybrid? Perhaps we construct a lazy precomputeddata every tick for the Movements to use, but pathing uses a fully computed one? Perhaps we check how big `Block.BLOCK_STATE_IDS.size()` is, and go lazy if it's above 10k otherwise precompute? Perhaps we benchmark how long precomputing takes, and do lazy if it's above 15ms, and precomputed on pathing thread if it's below 15ms?

Reply:
Which of these would you prefer I do. I would go for the simplest, just starting a separate thread, would this be alright, or do  you want me to go for one of these more complicated solutions?

Reply:
I'd really like detecting changes (by pointer equality, IMO setting values should be considered immutable) because it allows reusing data for multiple paths instead of recaching every time even though normally nothing changes. However I can understand if that's too much effort and in that case vote for a lazy cache using a byte/short/int/long per state because that should be comparable to an eager cache (also with a single value per state) in performance but without the setup overhead (literally just one extra branch on the special case path, no change on the fast path).
From my thoughts a hybrid lazy + non-lazy approach might ruin inlining and perform worse in all cases, so unless you show a benchmark proving that's wrong I'll vote against it.

Reply:
I think the first one is good for now. It avoids all issues of threading, precomputing, scheduling, all of that. Simply put in a `boolean[] blockStateHasBeenEvaluatedYet` or something. At the top of every query, check if it's false, if so we compute all the data (canWalkOn, canWalkThrough, etc) for this blockstate. It'll be fast because that check will almost always do nothing, so the branch predictor will have fun.

Reply:
It also means that `SettingChangedEvent` is not needed, but maybe we should leave it in, just in case @ZacSharp wants to write a setting changed detector? 😉

Reply:
Or I guess, for simplicity, it would be okay if the laziness was per-attribute. i.e. `canWalkOn(Blocks.STONE)` is cached unrelatedly to `canWalkThrough(Blocks.STONE)`. That way it could still be separate `PrecomputedDataForBlockState` objects.